platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/disk/OnDiskWritableSingletonState.java

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.state.merkle.disk;
+
+import com.hedera.pbj.runtime.Codec;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.state.spi.WritableSingletonStateBase;
+import com.swirlds.virtualmap.VirtualMap;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Objects;
+
+import static com.swirlds.state.merkle.logging.StateLogger.logSingletonRead;
+import static com.swirlds.state.merkle.logging.StateLogger.logSingletonRemove;
+import static com.swirlds.state.merkle.logging.StateLogger.logSingletonWrite;
+import static java.util.Objects.requireNonNull;
+
+public class OnDiskWritableSingletonState<T> extends WritableSingletonStateBase<T> {
+
+    @NonNull
+    private final VirtualMap megaMap;
+
+    @NonNull
+    private final Codec<T> valueCodec;
+
+    public OnDiskWritableSingletonState(
+            @NonNull final String serviceName,
+            @NonNull final String stateKey,
+            @NonNull final Codec<T> valueCodec,
+            @NonNull final VirtualMap megaMap) {
+        super(serviceName, stateKey);
+
+        this.valueCodec = requireNonNull(valueCodec);
+        this.megaMap = Objects.requireNonNull(megaMap);
+    }
+
+    // TODO: refactor? is is duplicated in OnDiskReadableSingletonState
+    /** {@inheritDoc} */
+    @Override
+    protected T readFromDataSource() {
+        final var value = megaMap.get(getMegaMapKey(), valueCodec);
+        // Log to transaction state log, what was read
+        logSingletonRead(getLabel(), value);
+        return value;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected void putIntoDataSource(@NonNull T value) {
+        megaMap.put(getMegaMapKey(), value, valueCodec);
+        // Log to transaction state log, what was put
+        logSingletonWrite(getLabel(), value);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected void removeFromDataSource() {
+        final var removed = megaMap.remove(getMegaMapKey(), valueCodec);
+        // Log to transaction state log, what was removed
+        logSingletonRemove(getLabel(), removed);
+    }
+
+    // TODO: refactor? is is duplicated in OnDiskReadableSingletonState
+    // TODO: test this method
+    /**
+     * Generates a 2-byte big-endian key identifying this singleton state in the Mega Map.
+     * <p>
+     * The underlying state ID (unsigned 16-bit) must be in [0..65535], and is written in big-endian order.
+     * </p>
+     *
+     * @return a {@link Bytes} object containing exactly 2 bytes in big-endian order
+     * @throws IllegalArgumentException if the state ID is outside [0..65535]
+     */
+    private Bytes getMegaMapKey() {
```

Comment:
It would be great to share such methods between readable and writable states (singleton, queue, and KV). For example, it can be defined in `ReadableSingletonStateBase`, which is then extended by `WritableSingletonStateBase` and `OnDiskReadableSingletonState`

Reply:
I thought about it and came to the conclusion that this method is an implementation detail of "OnDisk" classes, so I decided to find another way after addressing fundamental review points 🙂

Reply:
Optimized this